File: src/examples/python/runtime_adapter.py
import json

from axllm import AxCodeRuntime, AxCodeSession, RuntimeEnvelope, agent


class DemoSession(AxCodeSession):
    def __init__(self, globals, options=None):
        self.globals = dict(globals or {})
        self.create_options = dict(options or {})
        self.closed = False

    def execute(self, code, options=None):
        if "reservedNames" not in (options or {}):
            raise RuntimeError("reservedNames were not passed to the runtime")
        if code == "timeout()":
            return RuntimeEnvelope.timeout("demo timeout")
        self.globals["answer"] = "runtime final"
        return RuntimeEnvelope.final({"answer": self.globals["answer"]})

    def inspect_globals(self, options=None):
        return dict(self.globals)

    def snapshot_globals(self, options=None):
        return {"version": 1, "bindings": dict(self.globals), "closed": self.closed}

    def patch_globals(self, snapshot, options=None):
        self.globals = dict((snapshot or {}).get("bindings") or {})
        return self.snapshot_globals(options)

    def close(self):
        self.closed = True
        return {"closed": True}


class DemoRuntime(AxCodeRuntime):
    language = "Python"

    def __init__(self):
        self.sessions = []

    def create_session(self, globals, options=None):
        session = DemoSession(globals, options)
        self.sessions.append(session)
        return session


runtime = DemoRuntime()
runner = agent("question:string -> answer:string", {"runtime": {"language": "Python"}})
step = runner.execute_actor_step(runtime, "final()", {"question": "adapter"})
snapshot = runner.export_session_state()
runner.restore_session_state(snapshot)
timeout = runner.execute_actor_step(runtime, "timeout()", {"question": "adapter"})
closed = runner.close_runtime_session()

bindings = snapshot.get("bindings") or {}
print(
    json.dumps(
        {
            "stepKind": step["kind"],
            "finalArgs": step["completion_payload"]["args"],
            "snapshotKeys": sorted(bindings.keys()),
            "snapshotAnswer": bindings.get("answer"),
            "timeoutCategory": timeout["error_category"],
            "closed": closed,
        },
        indent=2,
        sort_keys=True,
    )
)

File: src/examples/cpp/runtime_adapter.cpp
#include "axllm/axllm.hpp"

#include <iostream>
#include <memory>
#include <vector>

struct DemoSession : axllm::AxCodeSession {
  axllm::Value globals;
  bool closed = false;

  explicit DemoSession(axllm::Value globals_) : globals(std::move(globals_)) {}

  axllm::Value execute(axllm::Value code, axllm::Value options = axllm::Value::object()) override {
    if (!axllm::Core::truthy(axllm::Core::map_contains(options, "reservedNames"))) {
      throw axllm::AxError("fixture", "reservedNames were not passed to the runtime");
    }
    if (axllm::equal(code, "timeout()")) return axllm::RuntimeEnvelope::timeout("demo timeout");
    axllm::Core::set(globals, "answer", "runtime final");
    return axllm::RuntimeEnvelope::final_payload({axllm::object({{"answer", axllm::Core::get(globals, "answer")}})});
  }

  axllm::Value inspect(axllm::Value = axllm::Value::object()) override { return globals; }

  axllm::Value snapshot_globals(axllm::Value = axllm::Value::object()) override {
    return axllm::object({{"version", 1}, {"bindings", globals}, {"closed", closed}});
  }

  axllm::Value patch_globals(axllm::Value snapshot, axllm::Value options = axllm::Value::object()) override {
    globals = axllm::Core::get(snapshot, "bindings", axllm::Value::object());
    return snapshot_globals(options);
  }

  axllm::Value close() override {
    closed = true;
    return axllm::object({{"closed", true}});
  }
};

struct DemoRuntime : axllm::AxCodeRuntime {
  std::vector<std::unique_ptr<DemoSession>> sessions;

  std::string language() const override { return "Python"; }

  axllm::AxCodeSession* create_session(axllm::Value globals, axllm::Value = axllm::Value::object()) override {
    sessions.push_back(std::make_unique<DemoSession>(std::move(globals)));
    return sessions.back().get();
  }
};

int main() {
  DemoRuntime runtime;
  auto runner = axllm::agent(
      "question:string -> answer:string",
      axllm::object({{"runtime", axllm::object({{"language", "Python"}})}}));
  axllm::Value step = runner.execute_actor_step(runtime, "final()", axllm::object({{"question", "adapter"}}));
  axllm::Value snapshot = runner.export_session_state();
  runner.restore_session_state(snapshot);
  axllm::Value timeout = runner.execute_actor_step(runtime, "timeout()", axllm::object({{"question", "adapter"}}));
  axllm::Value closed = runner.close_runtime_session();
  axllm::Value bindings = axllm::Core::get(snapshot, "bindings", axllm::Value::object());

  std::cout << axllm::stringify(axllm::object({
                   {"stepKind", axllm::Core::get(step, "kind")},
                   {"finalArgs", axllm::Core::get(axllm::Core::get(step, "completion_payload"), "args")},
                   {"snapshotKeys", axllm::Core::map_keys(bindings)},
                   {"snapshotAnswer", axllm::Core::get(bindings, "answer")},
                   {"timeoutCategory", axllm::Core::get(timeout, "error_category")},
                   {"closed", closed},
               }))
            << "\n";
}
